src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -180,6 +199,61 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
+                                googleProject: GoogleProject,
+                                clusterName: ClusterName,
+                                clusterRequest: ClusterRequest): Future[Cluster] = {
+    // Check if the google project has an active cluster with the same name. If not, we can create it
+    dbRef.inTransaction { dataAccess =>
+      dataAccess.clusterQuery.getActiveClusterByName(googleProject, clusterName)
+    } flatMap {
+      case Some(existingCluster) =>
+        throw ClusterAlreadyExistsException(googleProject, clusterName, existingCluster.status)
+      case None =>
+        val augmentedClusterRequest = addClusterDefaultLabels(
+          serviceAccountInfo, googleProject, clusterName, userEmail, clusterRequest)
+        val machineConfig = MachineConfigOps.create(clusterRequest.machineConfig, clusterDefaultsConfig)
+        val autopauseThreshold = calculateAutopauseThreshold(
+          clusterRequest.autopause, clusterRequest.autopauseThreshold)
+        val initialCluster = Cluster.create(
+          clusterRequest, userEmail, clusterName, googleProject,
+          serviceAccountInfo, machineConfig, dataprocConfig.clusterUrlBase, autopauseThreshold)
+
+        val attemptToSaveClusterInDb: Future[Cluster] = dbRef.inTransaction(_.clusterQuery.save(initialCluster))
+
+        // For the success case, register the following callbacks...
+        attemptToSaveClusterInDb foreach { cluster =>
+          // Notify the auth provider and kick off the cluster creation on the Google Dataproc side
+          asyncClusterCreate(userEmail, serviceAccountInfo, googleProject, clusterName, augmentedClusterRequest)
+            .onComplete {
+              case Success(_) =>
+                // Finally, notify the cluster monitor that the cluster has been created
+                clusterMonitorSupervisor ! ClusterCreated(cluster, clusterRequest.stopAfterCreation.getOrElse(false))
+              case Failure(_) =>
+                // If we didn't succeed, createGoogleCluster removes resources in Google but
+                // we also need to notify our auth provider that the cluster has been deleted.
+                // We won't wait for that deletion, though.
+                authProvider.notifyClusterDeleted(userEmail, userEmail, googleProject, clusterName)
+            }
+        }
+
+        attemptToSaveClusterInDb
+    }
+  }
+
+  def asyncClusterCreate(userEmail: WorkbenchEmail,
+                         serviceAccountInfo: ServiceAccountInfo,
+                         googleProject: GoogleProject,
+                         clusterName: ClusterName,
+                         clusterRequest: ClusterRequest): Future[(Cluster, GcsBucketName, Option[ServiceAccountKey])] = {
+    // Notify the AuthProvider
+    authProvider.notifyClusterCreated(userEmail, googleProject, clusterName)
+    // Proceed to creating the cluster on the Google side
+    // TODO Update this method so the stuff it returns is persisted
```

Comment:
So in the next PR it'll persist the Google stuff in the DB before kicking off the ClusterMonitorActor?

Reply:
Yes. That seems to be how it's done in the current version as well so I assume it would work. Let me know if I'm missing something.